// @chakra-ui/react-env/dist/esm/env.js
import React, { useMemo, useState, createContext, useContext } from "react";
import { ssrWindow } from "./mock-window";
import { ssrDocument } from "./mock-document";
var mockEnv = {
  window: ssrWindow,
  document: ssrDocument
};
var defaultEnv = typeof window !== "undefined" ? {
  window,
  document
} : mockEnv;
var EnvironmentContext = /*#__PURE__*/createContext(defaultEnv);
EnvironmentContext.displayName = "EnvironmentContext";
export function useEnvironment() {
  return useContext(EnvironmentContext);
}
export function EnvironmentProvider(props) {
  var {
    children,
    environment: environmentProp
  } = props;
  var [node, setNode] = useState(null);
  var context = useMemo(() => {
    var _ref;

    var doc = node == null ? void 0 : node.ownerDocument;
    var win = node == null ? void 0 : node.ownerDocument.defaultView;
    var nodeEnv = {
      document: doc,
      window: win
    };
    var env = (_ref = environmentProp != null ? environmentProp : nodeEnv) != null ? _ref : defaultEnv;
    return env;
  }, [node, environmentProp]);
  var shouldRenderChildren = node || environmentProp;
  return /*#__PURE__*/React.createElement(EnvironmentContext.Provider, {
    value: context
  }, shouldRenderChildren ? children : /*#__PURE__*/React.createElement("span", {
    ref: el => {
      if (el) setNode(el);
    }
  }));
}
